## G -Part

### Imports
*Initialize the environment and the Linear Regression class*

In [1]:
from lin_reg import LinearRegression
import numpy as np

datafile = "housing.csv"
model = LinearRegression()

np.set_printoptions(precision=2, suppress=True, linewidth=100)

data = np.genfromtxt(datafile, delimiter=",", names=True, dtype=None, 
                     encoding="utf-8",  missing_values="", filling_values=np.nan)

### Cleaning & Organizing
*Remove incomplete rows*

In [2]:
target_y = "median_house_value"
num_features = [
    "longitude", "latitude", "housing_median_age", 
    "total_rooms", "total_bedrooms", "population", 
    "households", "median_income"
]
cat_features = ["ocean_proximity"]
all_features = [target_y] + num_features + cat_features

mask = np.ones(len(data), dtype=bool)

for col in all_features:
    if data[col].dtype.kind in "fi":
        mask &= ~np.isnan(data[col])
    else: 
        mask &= (data[col] != "") & (data[col] != "nan")

clean_data = data[mask]

### Extra - Cleaning stats
*Summary of the data lost during the filtering process*

In [3]:
missing_results = []
for col in all_features:
    if data[col].dtype.kind in "fi":
        missing_count = np.sum(np.isnan(data[col]))
    else:
        missing_count = np.sum((data[col] == "") | (data[col] == "nan"))
    missing_results.append(f"{col}: {missing_count} missing values")
(missing_results, {
    "Total rows before": len(data),
    "Total rows remaining": len(clean_data),
    "Total rows lost": len(data) - len(clean_data)
})

(['median_house_value: 0 missing values',
  'longitude: 0 missing values',
  'latitude: 0 missing values',
  'housing_median_age: 0 missing values',
  'total_rooms: 0 missing values',
  'total_bedrooms: 207 missing values',
  'population: 0 missing values',
  'households: 0 missing values',
  'median_income: 0 missing values',
  'ocean_proximity: 0 missing values'],
 {'Total rows before': 20640,
  'Total rows remaining': 20433,
  'Total rows lost': 207})

### Categorical Features
*Transform text-based location data into binary numerical features*

In [4]:
unique_cats = np.unique(clean_data["ocean_proximity"])

cats_to_encode = unique_cats[:-1]
encoded_list = []

for cat in cats_to_encode:
    binary_col = np.zeros(len(clean_data))
    
    for i in range(len(clean_data)):
        if clean_data["ocean_proximity"][i] == cat:
            binary_col[i] = 1
    
    encoded_list.append(binary_col)

ocean_encoded = np.column_stack(encoded_list)
ocean_encoded.shape

(20433, 4)

### Preparing the data for the model
*Constructing the feature matrix X and the target vector y*

In [5]:
y = clean_data[target_y]
num_cols = []
for feature in num_features:
    num_cols.append(clean_data[feature])
X_num = np.column_stack(num_cols)

X_all = np.hstack([X_num, ocean_encoded])
intercept = np.ones(len(y))
X = np.column_stack([intercept, X_all])

### Train and use the model
*Executing the OLS fit and calculating standard error metrics*

In [6]:
model.fit(X,y)

variance = model.variance_calc(X,y)
std_dev = model.st_dev_calc(X,y)
rmse = model.rmse_calc(X,y)

[
    f"n (sample size): {model.n}",
    f"d (features): {model.d}",
    f"Variance (σ²): {variance:.2f}",
    f"Standard Deviation: {std_dev:.2f}",
    f"RMSE: {rmse:.2f}"
]


['n (sample size): 20433',
 'd (features): 12',
 'Variance (σ²): 4713776929.50',
 'Standard Deviation: 68656.95',
 'RMSE: 68635.11']

## VG Part

### $R^2$
*Evaluating how much of the variation in house prices the model explains.*

In [7]:
r_squared = model.r_squared_calc(X,y)
f"R-squared: {r_squared:.4f}"

'R-squared: 0.6465'

### F-Statistic & p-value
*Testing if the group of predictors together helps predict price better than using no predictors.*

In [8]:
f_stat, p_value = model.f_test(X,y)
{
    "F-Statistic": round(float(f_stat), 2),
    "p-value": f"{float(p_value):.10f}"
}

{'F-Statistic': 3111.61, 'p-value': '0.0000000000'}

### Pearson Correlation
*Analyzing linear relationships between feature pairs*

In [9]:
pearson = model.pearson_corr(X)
feature_names = ['intercept', 'longitude', 'latitude', 'housing_median_age', 
                 'total_rooms', 'total_bedrooms', 'population', 'households',
                 'median_income', 'ocean_<1H', 'ocean_INLAND', 'ocean_ISLAND', 
                 'ocean_NEAR_BAY']
short_names = ["LON", "LAT", "AGE", "RMS", "BED", "POP", "HHD", "INC", "<1H", "INL", "ISL", "BAY"]
rows = [f"{feature_names[i+1]:20s}: {pearson[i]}" for i in range(len(pearson))]

footer = " " * 24 + "   ".join(short_names)
rows.append(footer)
rows

['longitude           : [ 1.   -0.92 -0.11  0.05  0.07  0.1   0.06 -0.02  0.32 -0.06  0.01 -0.47]',
 'latitude            : [-0.92  1.    0.01 -0.04 -0.07 -0.11 -0.07 -0.08 -0.45  0.35 -0.02  0.36]',
 'housing_median_age  : [-0.11  0.01  1.   -0.36 -0.32 -0.3  -0.3  -0.12  0.05 -0.24  0.02  0.26]',
 'total_rooms         : [ 0.05 -0.04 -0.36  1.    0.93  0.86  0.92  0.2  -0.    0.03 -0.01 -0.02]',
 'total_bedrooms      : [ 0.07 -0.07 -0.32  0.93  1.    0.88  0.98 -0.01  0.02 -0.01 -0.   -0.02]',
 'population          : [ 0.1  -0.11 -0.3   0.86  0.88  1.    0.91  0.01  0.07 -0.02 -0.01 -0.06]',
 'households          : [ 0.06 -0.07 -0.3   0.92  0.98  0.91  1.    0.01  0.04 -0.04 -0.01 -0.01]',
 'median_income       : [-0.02 -0.08 -0.12  0.2  -0.01  0.01  0.01  1.    0.17 -0.24 -0.01  0.06]',
 'ocean_<1H           : [ 0.32 -0.45  0.05 -0.    0.02  0.07  0.04  0.17  1.   -0.61 -0.01 -0.31]',
 'ocean_INLAND        : [-0.06  0.35 -0.24  0.03 -0.01 -0.02 -0.04 -0.24 -0.61  1.   -0.01 -0.24]',


### Discussion : Collinearity and parameters selection
The Pearson correlation table shows that several predictors share very similar information. The strongest group is formed by the housing related variables: **total_bedrooms** correlates  with **households** at 0.98,**total_rooms** with **total_bedrooms** at 0.93, and with **households** at 0.92. These values are extremely high, which means these variables move almost in lockstep. **Population** is also closely tied to this group, with correlations above 0.85 with all 3 of them. Together, these four variables describe nearly the same underlying pattern in the dataset.

There is also a very strong relationship between the geographic variables longitude and latitude (−0.92). This indicates that the locations in the dataset follow a narrow diagonal pattern, so the two coordinates do not vary independently.

**Statistical Impact:**
When predictors are highly correlated, the variance-covariance matrix c = (X^T X)^(-1)σ² 
has inflated diagonal elements (cᵢᵢ). Since SE(B̂ᵢ) = σ̂ √cᵢᵢ, this directly increases 
standard errors. The model struggles to solve (X^T X)b = X^T y reliably when X^T X is 
near-singular due to collinearity.

Because these predictors overlap so much, the model cannot fully separate their individual effects. When several variables carry almost the same information, the coefficient of one variable can easily change sign depending on which of the others are included. This explains why total_rooms appears with the “wrong” sign: the model is trying to distribute a shared signal across several nearly identical predictors, and the sign becomes unstable as a result. This behaviour is typical in the presence of strong collinearity and does not mean that the variable has a genuinely negative relationship with the outcome.

These correlations do not create a statistical hierarchy in the regression sense. A statistical hierarchy would require keeping certain variables because they appear inside interaction terms or polynomial terms. Here, the predictors are correlated, but none of them are mathematically built from the others. The correlations therefore do not force any variable to remain in the model for hierarchical reasons.
Statistical hierarchy applies when:
- Interaction: Y = β₀ + β₁X₁ + β₂X₂ + β₃X₁X₂ (must keep X₁, X₂ if X₁X₂ is significant)
- Polynomial: Y = β₀ + β₁X + β₂X² (must keep X if X² is significant)

Our model has only raw variables, so no hierarchical constraint applies.
The housing variables do reflect different levels of the same structure which explains why they correlate so strongly. This structure justifies including them initially if the goal is to describe the housing environment in detail. However, their strong overlap limits how separately their coefficients can be interpreted. The unusual sign of total_rooms is a direct consequence of this overlap rather than a meaningful effect.

In short, the Pearson correlations show strong collinearity but no statistical hierarchy that forces us to keep variables in the model. The housing variables form a tightly connected cluster, which justifies their initial inclusion, but their overlap explains both the instability of coefficients and the unexpected sign of total_rooms. If the goal is clearer and more stable effects, simplifying this group is appropriate; if the goal is descriptive detail, the variables can be kept, as long as their dependence is acknowledged.

### T-tests
*Testing whether a single predictor has a real effect on price (different from zero).*

In [10]:
t_stats, p_values = model.t_tests(X,y)
results =  []


for i in range(len(t_stats)):
    sig = "Significant" if p_values[i] <0.4 else "NOT significant"
    results.append(f"{feature_names[i]:20s}: {t_stats[i]:8.4f}, {p_values[i]:.6f}, {sig}")

results

['intercept           : -25.6395, 0.000000, Significant',
 'longitude           : -26.2963, 0.000000, Significant',
 'latitude            : -25.3629, 0.000000, Significant',
 'housing_median_age  :  24.4389, 0.000000, Significant',
 'total_rooms         :  -7.8250, 0.000000, Significant',
 'total_bedrooms      :  14.6402, 0.000000, Significant',
 'population          : -35.2824, 0.000000, Significant',
 'households          :   6.6589, 0.000000, Significant',
 'median_income       : 116.1510, 0.000000, Significant',
 'ocean_<1H           :  -2.7258, 0.006421, Significant',
 'ocean_INLAND        : -19.3633, 0.000000, Significant',
 'ocean_ISLAND        :   4.8327, 0.000001, Significant',
 'ocean_NEAR_BAY      :  -3.7835, 0.000155, Significant']

All predictors in the fitted model are statistically detectable at the 40% significance level (every predictor's p‑value < 0.4). The t‑tests confirm that we reject H₀: βᵢ = 0 for all predictors, meaning each has a detectable effect on house price when controlling for all other variables.

With a sample size of 20,433, even very small effects become statistically detectable, so the fact that all predictors reject H₀ doesn't necessarily mean they're all practically important - it only confirms that effects exist.

*Confidence intervals in the next cell confirm these results: none include zero, supporting that all predictors have detectable effects.*

### Confidence Intervals
*Gives a plausible range for the true effect of a predictor; if the range does not include zero, the effect is detectable.*

In [11]:
lower, upper = model.confidence_interval(X,y,alpha=0.4)
ci_table = {}

for i in range(len(feature_names)):
    ci_table[feature_names[i]] = {
        "Lower": round(lower[i], 2),
        "Upper": round(upper[i], 2)
    }

ci_table


{'intercept': {'Lower': -2340048.64, 'Upper': -2191303.32},
 'longitude': {'Lower': -27671.17, 'Upper': -25954.81},
 'latitude': {'Lower': -26327.78, 'Upper': -24636.59},
 'housing_median_age': {'Lower': 1035.58, 'Upper': 1109.46},
 'total_rooms': {'Lower': -6.86, 'Upper': -5.53},
 'total_bedrooms': {'Lower': 94.78, 'Upper': 106.34},
 'population': {'Lower': -38.87, 'Upper': -37.06},
 'households': {'Lower': 43.35, 'Upper': 55.89},
 'median_income': {'Lower': 38975.09, 'Upper': 39544.05},
 'ocean_<1H': {'Lower': -5599.11, 'Upper': -2957.16},
 'ocean_INLAND': {'Lower': -45455.91, 'Upper': -41668.96},
 'ocean_ISLAND': {'Lower': 122740.28, 'Upper': 174507.33},
 'ocean_NEAR_BAY': {'Lower': -10063.41, 'Upper': -6400.96}}

-------------------------------------------------------------------------------------------------------- 

### Discussion on confidence Levels:

With R² = 0.6465, the model explains approximately 64.65% of the variance in house prices. Rounding down conservatively, a 60% confidence level (α = 0.40) is appropriate here. Rounding down rather than up ensures we do not overestimate the model's actual explanatory power.

Looking at the resulting confidence intervals, a few predictors stand out as clearly influential. Median_income has a strongly positive interval (38975, 39544), showing that higher income is consistently associated with higher predicted prices. 
Both longitude and latitude have negative intervals, indicating a stable geographic relationship with price. 
Among the categorical variables, ocean_ISLAND has a large positive interval (122740, 174507), while ocean_INLAND shows a clear negative effect (-45456, -41669).
Other predictors show smaller but consistent intervals. 
Total_rooms and population both have narrow negative intervals, while total_bedrooms and households show modest positive intervals. 

As discussed in the collinearity section, these housing variables should be interpreted carefully due to their high correlations.
The width of each interval reflects how precisely the coefficient is estimated: narrower intervals indicate more certainty, while wider ones show more uncertainty. 

Since none of the intervals include zero, each predictor has a clear and consistent effect on house prices at this confidence level.

### Extra - Analyzing high correlations:
*Highest multicollinearity*

In [12]:
corr_matrix = model.pearson_corr(X)
actual_features = feature_names[1:]
high_corr_list = []

for i in range(len(actual_features)):
    for j in range(i + 1, len(actual_features)):
        correlation = corr_matrix[i, j]
        if abs(correlation) > 0.85:
            high_corr_list.append((actual_features[i], actual_features[j], correlation))

high_corr_list.sort(key=lambda x: abs(x[2]), reverse=True)

[f"{feat_a:14s} & {feat_b:18s}: {val:.2f}" for feat_a, feat_b, val in high_corr_list]

['total_bedrooms & households        : 0.98',
 'total_rooms    & total_bedrooms    : 0.93',
 'longitude      & latitude          : -0.92',
 'total_rooms    & households        : 0.92',
 'population     & households        : 0.91',
 'total_bedrooms & population        : 0.88',
 'total_rooms    & population        : 0.86']

### Extra - Coefficients Magnitude

*How much each coefficient affects the price variance*

In [19]:
coef_list = []
for i in range(1, len(feature_names)):  # Skip intercept
    coef_list.append((feature_names[i], model.b[i]))

coef_list.sort(key=lambda x: abs(x[1]), reverse=True)

[f"{name:20s}={coef:12,.2f} $" for name, coef in coef_list]

['ocean_ISLAND        =  148,623.81 $',
 'ocean_INLAND        =  -43,562.43 $',
 'median_income       =   39,259.57 $',
 'longitude           =  -26,812.99 $',
 'latitude            =  -25,482.18 $',
 'ocean_NEAR_BAY      =   -8,232.19 $',
 'ocean_<1H           =   -4,278.13 $',
 'housing_median_age  =    1,072.52 $',
 'total_bedrooms      =      100.56 $',
 'households          =       49.62 $',
 'population          =      -37.97 $',
 'total_rooms         =       -6.19 $']

Looking at the absolute magnitude of coefficients provides insight into which predictors have the largest effects on house price in dollar terms.
However, it's important to separate continuous and categorical variables, as the first display a "per unit" effect while the others show the complete effect (since the values are actually 1 or 0, you can't have a "100 near bay"). 

**Continuous Variables (per unit effects):**

median_income has the largest coefficient (39,260), meaning each additional dollar of median income is associated with a $39,260 increase in predicted house price. 
Since median income varies by tens of thousands of dollars across neighborhoods, this translates to very large total price differences.
Geographic variables longitude (-26,813) and latitude (-25,482) show large per-degree effects, reflecting California's strong geographic price patterns where small location changes can mean large price differences.
housing_median_age (1,073) shows a moderate positive effect per year, while population (-38), households (50), total_bedrooms (101), and total_rooms (-6) show smaller per-unit effects.

**Categorical Variables (complete effects):**

The categorical ocean_proximity variables cannot be directly compared to continuous variables because they represent complete effects rather than per-unit changes. Each coefficient shows the full price difference of being in that location category:

ocean_ISLAND (148,624): Island properties are $148,624 more expensive
ocean_INLAND (-43,562): Inland properties are $43,562 cheaper
ocean_NEAR_BAY (-8,232): Near bay properties are $8,232 cheaper
ocean_<1H (-4,278): Properties within 1 hour of ocean are $4,278 cheaper

These are relative to the reference category (NEAR OCEAN, which was excluded during encoding).

### Extra - Refined Model demonstration

In [14]:
refined_indices = [0,1,2,3,4,6,8,9,10,11,12]
X_refined = X[:, refined_indices]

model_refined = LinearRegression()
model_refined.fit(X_refined, y)

[
 f"Unrefined RMSE: {rmse:.2f}" , 
 f"Unrefined Variance (σ²): {variance:.2f}",  
 f"Refined RMSE: {model_refined.rmse_calc(X_refined, y):.2f}", 
 f"Refined Variance (σ²): {model_refined.variance_calc(X_refined, y):.2f}"
 ]

['Unrefined RMSE: 68635.11',
 'Unrefined Variance (σ²): 4713776929.50',
 'Refined RMSE: 70690.98',
 'Refined Variance (σ²): 4999906984.29']

In [15]:
t_stats_ref, p_values_ref = model_refined.t_tests(X_refined, y)
ref_feature_names = [
    'intercept', 'longitude', 'latitude', 'housing_median_age', 
    'total_rooms', 'population', 'median_income', 
    'ocean_<1H', 'ocean_INLAND', 'ocean_ISLAND', 'ocean_NEAR_BAY'
]

original_results = ["--- ORIGINAL MODEL ---"]
for i in range(len(t_stats)):
        original_results.append(f"{feature_names[i]:20s}: {t_stats[i]:8.4f}, {p_values[i]:.6f}")
refined_results = ["--- REFINED MODEL ---"]
for i in range(len(t_stats_ref)):
        refined_results.append(f"{ref_feature_names[i]:20s}: {t_stats_ref[i]:8.4f}, {p_values_ref[i]:.6f}")

comparison = original_results + [""] + refined_results
comparison

['--- ORIGINAL MODEL ---',
 'intercept           : -25.6395, 0.000000',
 'longitude           : -26.2963, 0.000000',
 'latitude            : -25.3629, 0.000000',
 'housing_median_age  :  24.4389, 0.000000',
 'total_rooms         :  -7.8250, 0.000000',
 'total_bedrooms      :  14.6402, 0.000000',
 'population          : -35.2824, 0.000000',
 'households          :   6.6589, 0.000000',
 'median_income       : 116.1510, 0.000000',
 'ocean_<1H           :  -2.7258, 0.006421',
 'ocean_INLAND        : -19.3633, 0.000000',
 'ocean_ISLAND        :   4.8327, 0.000001',
 'ocean_NEAR_BAY      :  -3.7835, 0.000155',
 '',
 '--- REFINED MODEL ---',
 'intercept           : -25.9566, 0.000000',
 'longitude           : -27.0413, 0.000000',
 'latitude            : -26.2918, 0.000000',
 'housing_median_age  :  22.2453, 0.000000',
 'total_rooms         :  30.6387, 0.000000',
 'population          : -27.5044, 0.000000',
 'median_income       : 111.6947, 0.000000',
 'ocean_<1H           :  -2.4349, 0.014905

The refined model confirms that the unusual behaviour in the original regression was caused by multicollinearity. When several predictors describe almost the same underlying pattern, the model cannot assign their effects independently, and the coefficients become unstable. This is why total_rooms appeared with a negative sign despite representing a feature that should logically increase housing value. After removing the most overlapping variables (total_bedrooms and households), the remaining predictors no longer compete for the same signal, and total_rooms returns to a positive and more interpretable effect.

The refined model has a slightly higher RMSE and variance, which means its predictions are less precise. This happens because removing redundant predictors reduces the model’s ability to “fit” the sample perfectly. However, the trade‑off is beneficial for interpretation. The refined model produces coefficients that are more stable, easier to interpret, and less influenced by noise introduced by collinearity. For understanding how each predictor relates to housing prices, the refined model is therefore the more trustworthy one, while the full model remains better only for raw predictive accuracy.
